Block API: Allow block registration without category

aduth opened the pull request:
Alternative to: #22192

This pull request seeks to update block registration to allow for the omission of a category in block settings. This differs from #22192 in that it avoids treating "Uncategorized" as a formalized, default category for blocks. Instead, it simply allows blocks to lack a `category` property value. The "downside" here is that it then relies on each individual UI of the editor to account for the potential absence of category. Currently, this appears to affect only the Inserter and the Block Manager components. Note that changes would be required in #22192 as well for these components, so there's not a significant difference between the approaches in this regard.

~Status: In Progress. Remaining tasks:~

- [x] Update Block Manager to include "Uncategorized" section.
   - **Edit:** Updated in 043ebc5.
- [x] Validate or incorporate fix for missing category in search ([see change from #22192](https://github.com/WordPress/gutenberg/pull/22192/commits/a28d026b51c23a0327aa3a3000323e8386611126#diff-4e302c362c07ad0b65925c3de0b9bec3L290-R301))
   - **Edit:** Added a test case in bf5f678. The implementation appears to handle it okay. The test case should at least help assure there's no regressions in the future.

**Testing Instructions:**

To test, you need to register a block type which does not have an assigned category, or which has an invalid category. The easiest way I found to do this was to locally modify the settings of a core block, e.g. [the paragraph block](https://github.com/WordPress/gutenberg/blob/master/packages/block-library/src/paragraph/block.json), to remove or alter the "category" field.

Then, verify that the block is shown under the "Uncategorized" categorizations within the block inserter and Block Manager modal.

*Inserter:*

1. Navigate to Posts > Add New
2. Toggle the block inserter
3. Observe block under "Uncategorized"
4. Confirm that you can still search for the block

*Block Manager:*

1. Navigate to Posts > Add New
2. Toggle the "More Tools & Options" menu in the top-right corner
3. Click "Block Manager"
4. Observe block under "Uncategorized"
5. Confirm that you can still search for the block

Ensure unit tests pass:

```
npm run test-unit
```

aduth committed:
Block API: Allow block registration without category

aduth committed:
Block Editor: Show uncategorized block items in inserter

aduth commented on packages/block-editor/src/components/inserter/block-list.js:
```diff
@@ -135,6 +135,10 @@ function InserterBlockList( {
 		return filter( filteredItems, { category: 'reusable' } );
 	}, [ filteredItems ] );
 
+	const uncategorizedItems = useMemo( () => {
```
I think the pattern we've arrived at with various filtering on blocks has grown to a point where it could probably do for a refactoring to generalize the approach. I wasn't very keen to do it here, since it would inevitably touch on a few other problems with the implementation very much outside of scope (#22279, https://github.com/WordPress/gutenberg/pull/17609#issuecomment-626938994, https://github.com/WordPress/gutenberg/pull/17609#issuecomment-626944735). In the interim, I've merely adopted the established pattern.

youknowriad replied:
Should we include also the items that have a "category" but that category is not registered?

aduth replied:
> Should we include also the items that have a "category" but that category is not registered?

It works this way (indirectly) as a result of the additional changes in `registration.js`, where the `settings.category` is unset if it had been set to a category which is not registered.

youknowriad replied:
The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

aduth replied:
> The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

Maybe it's something where the `blockTypes` reducer should handle this behavior, including in response to category-affecting action types? I guess the downside is that if an earlier registered block type category had been considered invalid but is now _valid_, it wouldn't be restored.

Another option could be that this is handled somewhere in a selector like `getBlockType`.

In the end, I think it depends if we'd rather have certainty in knowing that data "at rest" is valid (or at least valid from the selector access point), or if we want to push this onto the specific UI to handle. The latter seems it could be cumbersome in how essentially it would need access to all categories just to determine if the category for a particular block type is valid.

youknowriad replied:
Since these selectors don't change that often (unlikely to cause performance issues), I personally think the latter approach is probably the best one. Memoization does have some costs though.

aduth replied:
> The issue is that one can trigger `setCategories` after the block has been registered since it's just a redux action.

I also wonder how much of a problem this is in practice, and whether it's something where we can be content in saying "the value is validated at the time of registration", irrespective of what happens after that point in time. In a sense, this is consistent with how block registration works in other ways. We currently forbid a block from being registered if its category doesn't exist. We never previously cared if that category _came into existence_ after that attempt at registration.

I'm fine any of these ways, to be honest. The selector seems like the best middle-ground, but as you note, there's some trade-offs to consider here, and a slight extra complexity in needing to account for this outside of registration-time considerations.

youknowriad replied:
I agree that it's not a big concern and I'm happy to move forward as is in this PR.

I'm thinking more as an ideal world here:

 - no usage of `wp.blocks.*` APIs.
 - rely 100% on selectors and actions.
 - registration timing not impactful.

There's a larger goal/dream here and just trying to see how we can fit it here :).

aduth replied:
Yeah, it sounds nice. I started to consider another alternative to the selector, which is to have a second "internal" state which maintains all initial block registrations, irrespective of their validity. It would be useful for categories, to solve what I described in https://github.com/WordPress/gutenberg/pull/22280#discussion_r423734129 as the problem of "an earlier registered block type category had been considered invalid but is now valid, it wouldn't be restored". I could see how something like this could be useful for other types of validations which can change over time. Looking at the [registration logic](https://github.com/WordPress/gutenberg/blob/89e21cf632a352c38844e12a34cec22bfefd12a6/packages/blocks/src/api/registration.js#L142-L242), I'm not sure anything else would fall under this.

I'd probably agree it's something we could consider addressing separately.

ZebulanStanphill commented on packages/block-editor/src/components/inserter/block-list.js:
```diff
@@ -135,6 +135,10 @@ function InserterBlockList( {
 	}, [ filteredItems ] );
 
+	const uncategorizedItems = useMemo( () => {
+		return filter( filteredItems, ( item ) => ! item.category );
```
```suggestion
		return filteredItems.filter( ( item ) => ! item.category );
```

Or if `filteredItems` can be `nullish`:

```suggestion
		return filteredItems?.filter( ( item ) => ! item.category );
```

aduth replied:
I was reluctant at first, since for consistency's sake I was merely following the established form. But I think it's fair enough to do a refactoring pass here, since it's likely not to be done for its own sake, and the pull request isn't too overwhelming as it is.

Updated in 90b6470.

youknowriad commented on packages/block-editor/src/components/inserter/block-list.js:
```diff
@@ -228,6 +235,19 @@ function InserterBlockList( {
+			{ ! hasChildItems && uncategorizedItems.length && (
+				<InserterPanel
+					className="block-editor-inserter__uncategorized-blocks-panel"
+					title={ __( 'Uncategorized' ) }
```
For uncategorized patterns, I've used "All" as a label. "Uncategorized" is too technical and can confuse users. "All" is inaccurate. Not sure what's best.

aduth replied:
> For uncategorized patterns, I've used "All" as a label. "Uncategorized" is too technical and can confuse users. "All" is inaccurate. Not sure what's best.

Well, there is the precedent of the "Uncategorized" default post category 🤷

ZebulanStanphill replied:
If we don't just go with "Uncategorized", then I think "Other" would work.

aduth replied:
I think it should be fine to stick with "Uncategorized"? I guess we can at least see what feedback there is, if any.

aduth committed:
Blocks: Add test case to verify behavior omitting category

Search implementation leverages block category, which may be absent. It did not actually require changes to the implementation to support, but the test case can help assure this continues to be supported into the future.

aduth committed:
Edit Post: Display uncategorized blocks as section in Manage Blocks

aduth committed:
Block Editor: Replace Lodash with native array filter

aduth committed:
Block Editor: Inserter: Avoid zero output in rendering

aduth commented:
>Update Block Manager to include "Uncategorized" section.

Updated in 043ebc5.

> * Validate or incorporate fix for missing category in search ([see change from #22192](https://github.com/WordPress/gutenberg/pull/22192/commits/a28d026b51c23a0327aa3a3000323e8386611126#diff-4e302c362c07ad0b65925c3de0b9bec3L290-R301))

Added a test case in bf5f678. The implementation appears to handle it okay. The test case should at least help assure there's no regressions in the future.

gziolo commented:
I added the `Needs Dev Note` label, but it really should be combined/added to the one proposed in https://github.com/WordPress/gutenberg/pull/19279#issuecomment-635982767.
